x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/GroupingAggregatorFunction.java

```diff
@@ -48,12 +53,40 @@ interface AddInput extends Releasable {
          *     be skipped entirely or the groupIds block could contain a
          *     {@code null} value at that position.
          * </p>
+         * <p>
+         *     This method delegates the processing to the other overloads for specific groupIds block types.
+         * </p>
          * @param positionOffset offset into the {@link Page} used to build this
          *                       {@link AddInput} of these ids
          * @param groupIds {@link Block} of group id, some of which may be null
          *                 or multivalued
          */
-        void add(int positionOffset, IntBlock groupIds);
+        default void add(int positionOffset, IntBlock groupIds) {
+            switch (groupIds) {
+                case ConstantNullBlock ignored:
+                    // No-op
+                    break;
+                case IntVectorBlock b:
+                    add(positionOffset, b.asVector());
+                    break;
+                case IntArrayBlock b:
+                    add(positionOffset, b);
+                    break;
+                case IntBigArrayBlock b:
+                    add(positionOffset, b);
+                    break;
+            }
+        }
```

Comment:
This class and this method specifically are the core of the PR.

For this method, I would have liked to make it `final`, as to avoid overrides disabling the enhancement. But it's an interface, and making it a class is probably (maybe?) not worth it _(Insert more question marks here)_

As for the implementation, luckily, I could simply ignore the `ConstantNullBlock` impl, and delegate the `IntVectorBlock` one to the `add(IntVector)` existing method. So it looks better than I expected

Reply:
You could make it `static` maybe, but that'd involve changing a lot of call sites. I'd probably change it for a follow up you can call "100% mechanical"